randovania/gui/widgets/game_connection_window.py

```diff
@@ -165,12 +197,57 @@ def setup_builder_ui(self):
         # TODO: multiworld can't handle multiple builders, so disable it for now
         self.add_builder_button.setEnabled(not self.game_connection.connection_builders)
 
-        self.update_builder_ui()
+        self.game_connection.BuildersUpdated.emit()
+
+    @asyncSlot()
+    async def _attempt_join(self, builder: ConnectorBuilder):
+        if not await self.network_client.ensure_logged_in(self):
+            return
+        session_list_entry = self.session_for_builder[builder]
+        joined_session = await self.network_client.attempt_join(session_list_entry)
+        if joined_session is not None:
+            await self.window_manager.activate_or_create_game_session_window(
+                self.network_client, joined_session,
+                self.options
+            )
+
+    async def _try_to_find_session(self, builder: ConnectorBuilder) -> bool:
+        # check if there is a remote connector for the builder
+        associated_rc = self.game_connection.remote_connectors.get(builder, None)
+        if associated_rc is None:
+            return False
+        
+        # check the uuid
+        layout_uuid = associated_rc._layout_uuid
+        if layout_uuid == INVALID_UUID:
+            return False
+
+        if not await self.network_client.ensure_logged_in(self):
+            return False
+        session_list_entry = await self.network_client.get_session_for_uuid(layout_uuid)
```

Comment:
Hmmm you mean both calls? From the other comment, I get that we could remove L225-226 since you can just sync it without being logged in.
L227 is just to make sure that the UUID really belongs to a session. It could be also removed and just be like "not invalid = show button" and do the login when the user really pressed the button to open the session window.

Reply:
After working on it locally: The list entry, which is holding the fields for "has password" and "user in session", is not accessible by `WorldDatabase` (or I just don't find it). 
That means I just delay the calls until the user presses the join action? Do a login, ask for session entry, optional password prompt and eventually open the session window.
I must admit asking for a session list entry is mostly needed because of rare corner cases like:
- Having a valid UUID but there is no session for it
- valid UUID + session exists but you are not in the session (and it has a password). That's like trying to steal someone's session by getting access to the world's iso / romfs.

Otherwise we could just login and allow to join / open the session window by just sending a uuid and the server does the rest and just declines it if something is wrong.

Reply:
My idea is that if the user knows about a World and even has session data about it, then 99% of the time [1] that World belongs to a Session the user also belongs to.
So if you're connected to a Game and you know session details about that Game's World, we should assume "user in session" is valid, and if that's valid we should be able to not need a password. This follows with:
- Call `multiplayer_listen_to_session` to make the server start sending you session details.
   - This request fails if you're not a member already.
- Call `multiplayer_request_session_update` to get a full entry, though looking at `MultiplayerSessionWindow.create_and_update` we could get away with just passing a session id here, especially since it already requests the update.

[1]: I can only think of "logged out and then logged in as another user" and "kicked out of session" as reasons why that wouldn't happen. Both are rare enough that just handling as failure case seems good enough.

> Having a valid UUID but there is no session for it

If WorldDatabase has session details for a world, then a session absolutely exists, and you have it's id!
And if doesn't (session got deleted? then the world itself also got deleted and other mess), you `multiplayer_listen_to_session` will fail because you're not a member :)

> valid UUID + session exists but you are not in the session

I would just reject completely if you're not a member. Put the session name in the error message if needed, but let the user join via the normal flow. This is a rare case, let's not overcomplicate it.

Reply:
Changed the logic again. Re-requested review but take your time. Earliest I will fix the tests is somewhere on the weekend.